### ISSUE TITLE
feat: gradient-aware global optimisers (hybrid/memetic + gradient-noise) to escape local minima

### ISSUE THREAD

**Issue by FinbarArgus** [physiomelinks/circulatory_autogen#252]
## Summary

Add **gradient-aware global optimisers** to the param-id backend — methods that exploit gradient information (which we *already* compute via CasADi AD) while still escaping local minima, unlike a pure local solver like `sp_minimize` (L-BFGS-B).

Today the backend (`src/param_id/optimisers.py`) offers four strategies that sit at opposite ends of a spectrum:

| `param_id_method` | Class | Global? | Uses gradients? |
|---|---|---|---|
| `genetic_algorithm` | `GeneticAlgorithmOptimiser` (custom, MPI) | ✅ global | ❌ |
| `CMA-ES` | `CMAESOptimiser` (nevergrad) | ✅ global | ❌ |
| `bayesian` | `BayesianOptimiser` (skopt, deprecated) | ✅ global | ❌ |
| `sp_minimize` | `SciPyMinimizeOptimiser` (scipy L-BFGS-B) | ❌ **local only** | ✅ AD via `get_jac_cost_ca` |

So we either explore globally **and throw the gradient away**, or use the gradient **and get stuck in the first local minimum**. The literature on hybrid/memetic and gradient-noise methods is exactly about closing that gap: pair global exploration (escape local optima) with gradient-based local exploitation (fast, accurate refinement). This issue catalogues the options and rates how hard each is to add **here**, given our architecture.

### Why this is cheap for us specifically
- **Analytic gradients already exist.** For `casadi_python` models, `param_id_obj.get_jac_cost_ca(p)` returns `dJ/dp` and `get_cost_ca(p)` the cost — this is what `SciPyMinimizeOptimiser` already consumes. Any gradient-hybrid method can reuse it directly. (For `cellml_only`/Myokit models gradients are finite-difference or pending — see #234 and #239.)
- **Plugging in a new optimiser is mechanical:** subclass `Optimiser` (ABC in `optimisers.py`), implement `run()` (set `self.best_param_vals` / `self.best_cost`, save via `_save_best_params()`), then add one `elif self.param_id_method == '<name>'` branch in `OpencorParamID.run_param_id` (`src/param_id/paramID.py:1397-1453`) and document the name in `user_inputs.yaml` / docs.
- `nevergrad` and `scikit-optimize` are **already dependencies**, and MPI parallel cost eval (`get_cost_from_params`) is already wired.

---

## Options (roughly easiest → hardest)

Difficulty is relative to *this* codebase (new `Optimiser` subclass + dispatch branch + test), not from scratch.

### 1. Sequential / memetic hybrid: global search → L-BFGS polish  🟢 Easy (~0.5–1 day)
**Idea (decoupled hybridisation):** run `CMA-ES` (or GA) to locate a promising basin globally, then hand the best point to the existing L-BFGS-B + AD `sp_minimize` for accurate local refinement. This is the simplest, most robust win and mirrors the `jSO_CMA-ES_LBFGS` family ([Zhao et al. 2022](https://doi.org/10.1007/s00521-021-06849-z)).
- **Implement:** a `HybridGlobalLocalOptimiser` that calls the two existing optimisers in sequence — almost no new numerics, both pieces already exist.
- **Libraries:** none new (reuses `nevergrad` + `scipy`).
- **Risk:** very low. Could even be a config flag (`polish_with_gradient: true`) on the existing global methods.

### 2. Expose nevergrad's built-in meta/hybrid optimisers  🟢 Easy (~0.5 day)
**Idea:** [`nevergrad`](https://facebookresearch.github.io/nevergrad/) already ships gradient-aware and portfolio hybrids — `NGOpt` (auto-selects), `Chaining([CMA, ...], ...)` to chain global→local, `MetaModel` (surrogate-assisted CMA), `CMandAS`/`BFGSCMAPlus`. We already wrap `ng.optimizers.CMA` in `CMAESOptimiser`.
- **Implement:** generalise `CMAESOptimiser` to take an `ng_optimizer_name` option (e.g. `param_id_method: nevergrad`, `optimiser_options: {ng_optimizer: NGOpt}`), passing through the existing ask/tell MPI loop.
- **Libraries:** [nevergrad](https://github.com/facebookresearch/nevergrad) (already a dep).
- **Risk:** very low; biggest value-to-effort ratio alongside #1.

### 3. Basin hopping  🟢🟡 Easy–Medium (~1–2 days)
**Idea:** repeated random "hops" in parameter space, each followed by a gradient local minimisation, accepting/rejecting by a Metropolis criterion — escapes local minima while exploiting gradients at every step.
- **Implement:** [`scipy.optimize.basinhopping`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.basinhopping.html) with `minimizer_kwargs={'method':'L-BFGS-B','jac': <AD gradient>}` reusing the `sp_minimize` gradient wrapper. Mostly serial (rank-0) unless we parallelise hops over MPI.
- **Libraries:** `scipy` (already a dep).
- **Risk:** low; main work is MPI/bounds/normalisation plumbing + history logging.

### 4. Multi-start + clustering global optimisation (MLSL)  🟡 Medium (~2–4 days)
**Idea:** scatter many starts, run a gradient local solver from each, and use clustering (Multi-Level Single Linkage) to avoid re-converging to the same minimum — a classic, theoretically grounded global method that fully uses gradients.
- **Implement:** wrap [`nlopt`](https://nlopt.readthedocs.io/en/latest/NLopt_Algorithms/#mlsl-multi-level-single-linkage) `GD_MLSL` / `GD_MLSL_LDS` (gradient local solver = LBFGS/MMA), feeding our AD gradient via NLopt's `f(x, grad)` callback. Embarrassingly parallel starts map naturally onto MPI.
- **Libraries:** [NLopt](https://github.com/stevengj/nlopt) (new dep, `pip install nlopt`).
- **Risk:** medium — new dependency + adapting the gradient/bounds to NLopt's callback signature.

### 5. Stochastic-gradient + noise injection (SGLD / noisy gradient descent)  🟡 Medium (~3–5 days)
**Idea:** gradient descent with properly-scaled Gaussian noise added each step provably escapes saddle points and shallow local minima ([Welling & Teh 2011 SGLD](https://www.stats.ox.ac.uk/~teh/research/compstats/WelTeh2011a.pdf); [Raginsky et al. 2017](https://proceedings.mlr.press/v65/raginsky17a.html); [Ge et al. 2015](https://proceedings.mlr.press/v40/Ge15.html)). Scales to high dimensions far better than population methods, and the Langevin variant doubles as a posterior sampler (synergy with the MCMC work, #195/#193).
- **Implement:** a custom loop over `get_jac_cost_ca` with a step-size/temperature schedule and noise term — straightforward numerics, but needs tuning of the noise/annealing schedule to be robust. Could prototype with [`torch.optim`](https://pytorch.org/docs/stable/optim.html) SGD + manual noise, or [`pyro`](https://docs.pyro.ai/)/[`numpyro`](https://num.pyro.ai/) for the full Bayesian SGLD/SGHMC.
- **Libraries:** none required (numpy + existing AD); optional torch/pyro.
- **Risk:** medium — most experimental; benefit strongest for high-dim / many-parameter models and uncertainty quantification.

### 6. Gradient-informed population methods (memetic GA / gradient-enhanced PSO)  🟠 Hard (~1–2 weeks, or Medium via pagmo)
**Idea:** inject gradient local search *inside* the population loop — e.g. memetic GA with adaptive gradient local search ([MA-ALS, Arab & Alfi 2015](https://doi.org/10.1016/j.ins.2014.11.051)), gradient-enhanced PSO ([GPSO/SAPSO](https://doi.org/10.1016/j.asoc.2018.04.021)), or diversity-maintained DE with embedded gradients (DMGBDE). Best final accuracy in principle, but most code.
- **Implement either:**
  - Custom: extend our MPI `GeneticAlgorithmOptimiser` to Lamarckian local-search a fraction of the population via `get_jac_cost_ca` each generation. Most flexible, most work.
  - Off-the-shelf: [`pygmo`/pagmo](https://esa.github.io/pygmo2/) provides ready memetic algorithms and a global+local archipelago model (Medium effort, heavyweight dep).
- **Libraries:** [pygmo](https://github.com/esa/pagmo2) (new, heavy) or none (custom).
- **Risk:** higher; recommend only after #1–#2 prove the gradient-polish pattern.

### 7. Differentiable metaheuristics (EvoGrad-style)  🔴 Research (weeks)
**Idea:** make the metaheuristic itself differentiable so loss gradients flow through selection/mutation/particle updates ([EvoGrad, arXiv:2506.06320, 2025](https://arxiv.org/abs/2506.06320)). Cutting-edge; listed for completeness, not recommended near-term — would need a full autodiff (torch/JAX) reframing of the optimiser.

---

## Recommendation

Start with **#1 (CMA-ES→L-BFGS polish)** and **#2 (nevergrad `NGOpt`/`Chaining`)** — both are ~half-day changes reusing existing code and dependencies, and together they deliver most of the "use the gradient without getting stuck" benefit. Add **#4 (MLSL)** or **#5 (SGLD)** afterward if we need stronger global guarantees or high-dimensional / UQ scaling.

## Acceptance criteria
- [ ] At least one new gradient-aware global `param_id_method` selectable via `user_inputs.yaml`.
- [ ] Reuses the existing AD gradient (`get_jac_cost_ca`) where available; documented fallback when it isn't.
- [ ] Works under MPI (`run_param_id.sh`) and writes the standard `best_cost` / `best_param_vals` + history CSVs.
- [ ] Test added in `tests/test_param_id.py` (use the `compare_optimisers` marker) showing it beats `sp_minimize` on a deliberately multi-modal fixture.

## References / related
- Related issues: #234 (gradient `sp_minimize` for cellml_only), #239 (CVODES forward sensitivities through Myokit), #195/#193 (MCMC — SGLD synergy).
- Survey context: gradient-metaheuristic hybridisation taxonomy ([Pujari & Veeramachaneni, arXiv:2312.09703](https://arxiv.org/abs/2312.09703)); CMA-ES variants & hybrids survey ([Archives of Comp. Methods in Eng. 2026](https://doi.org/10.1007/s11831-026-10557-z)).

<sub>Compiled from a fact-checked literature search (hybrid/memetic global-local optimisation + stochastic-gradient noise injection). Difficulty estimates are relative to the current `optimisers.py` / `paramID.py` plug-in pattern.</sub>

**Comment by lakshtanov** [comment]
@FinbarArgus — some thoughts on gradient-aware global optimisation for ODE model calibration.

## Part I: Mathematical Overview

### The problem

Minimise a cost functional J(p) where p is a parameter vector (dim 4-10), J involves solving an ODE system and comparing the trajectory to observations. The landscape has local minima due to nonlinearity of the model. An exact gradient ∇J is available at a cost comparable to one forward evaluation.

### The tension

- **Local gradient methods** (L-BFGS-B, Newton): converge in O(20) evaluations but get trapped in the nearest local minimum.
- **Global stochastic methods** (GA, CMA-ES): explore the full landscape but require O(1000-50000) evaluations, ignoring gradient information entirely.

The goal is to combine global exploration with gradient-based local exploitation.

### Method 1: Multi-start local optimisation

Sample N points from the parameter domain, run a local gradient solver from each, keep the best result. If the number of basins of attraction is finite, this finds the global minimum with probability approaching 1 as N grows.

- For physics-constrained models the number of distinct basins grows slowly with dimension; O(50-100) starts typically suffice for 4-10 parameters.
- **MLSL variant** (Multi-Level Single Linkage, Rinnooy Kan & Timmer 1987): clusters converged points to avoid redundant local solves.

### Method 2: Basin-hopping (Wales & Doye, 1997)

Transform the cost landscape: map each point to the minimum of its basin, then explore this simplified landscape via random perturbation + Metropolis acceptance.

- Effective when basins have similar depth but different widths.
- Requires tuning the perturbation scale (problem-specific).

### Method 3: Langevin dynamics

Gradient descent with calibrated noise:

dX = -∇J dt + σ(t) dW

- **Annealing** (σ(t) → 0): provably converges to the global minimum under mild conditions (Gelfand & Mitter, 1991).
- **Constant temperature**: samples from the Boltzmann distribution exp(-J/σ²), giving a posterior distribution over parameters rather than a point estimate.
- **SGLD** (Welling & Teh, 2011): stochastic gradient variant — mini-batch noise replaces explicit noise injection. Provably explores multi-modal landscapes.
- **HMC** (Hamiltonian Monte Carlo): uses gradient to construct energy-preserving trajectories, giving efficient posterior sampling. The Bayesian extension of Langevin dynamics.
- Theoretical convergence guarantees: Raginsky, Rakhlin & Telgarsky (2017).

### Method 4: Mini-batch stochastic gradient (the "training" analogy)

In neural network training, optimisers like Adam escape local minima because mini-batch sampling creates stochastic gradients. For ODE calibration the loss is deterministic — but we can create noise artificially:

**Sub-sample the observation time points.** Instead of computing cost over all T observations, use a random subset of size B < T at each step. This produces gradient noise proportional to √(T/B - 1). The noise is automatically scaled by the landscape: large where observations disagree (rough landscape), small where they agree (smooth landscape).

Increasing B over time (batch-size annealing) transitions from global exploration to local convergence. A final polish with the full dataset gives precise convergence.

This is equivalent to SGLD (Method 3) but with a natural, problem-adapted noise source instead of explicit Gaussian injection.

### Method 5: Multiple shooting (Bock & Plitt, 1984)

Specific to ODE models. Split the trajectory into K segments. Introduce the initial conditions at each segment boundary as additional optimisation variables. Minimise:

J(p, y₁, ..., y_K) = data fit + Σ ‖continuity gaps‖²

This addresses the **source** of local minima: in single-shooting, small parameter changes early in the trajectory cascade into large trajectory changes later, creating an artificially rough landscape. Multiple shooting decouples segments, making the landscape smoother and the local minima shallower or absent.

### Method 6: Population + gradient (memetic algorithms)

Inject gradient-based local search inside a population-based loop: memetic GA (gradient polish of selected individuals), gradient-enhanced PSO, Natural Evolution Strategies (NES — gradient of expected fitness). Most powerful in theory, most complex to implement.

### Summary table

| Method | Uses gradient | Global | Tuning needed | Complexity |
|--------|:---:|:---:|---|---|
| L-BFGS-B (local) | Yes | No | None | Trivial |
| Multi-start local | Yes | Yes (probabilistic) | N (number of starts) | Trivial |
| Basin-hopping | Yes | Yes | Perturbation scale | Low |
| Langevin / SGLD | Yes | Yes (provable) | Temperature schedule | Low-Medium |
| Mini-batch SGD/Adam | Yes | Yes | Batch size schedule | Low-Medium |
| Multiple shooting | Yes | Indirect (smoother landscape) | Number of segments | Medium |
| Memetic population | Yes | Yes | Population size, selection | High |

---

## Part II: Our Experience and Proposals

### What we have tested

On the 3-compartment cardiovascular model (27 states, 4 calibration parameters):
- **L-BFGS-B with AD gradient**: 0.2 seconds for full calibration. Gradient evaluation ~6ms.
- **HMC posterior sampling**: 200 samples in 18 seconds — Langevin dynamics (Method 3) taken to its Bayesian conclusion.

The gradient cost was negligible compared to the forward ODE solve. This makes all gradient-based methods (1-4) practical.

### Proposal A: Multi-start L-BFGS-B

L-BFGS-B alone is a local optimiser: it converges fast (~20 gradient evaluations) but gets trapped in the nearest local minimum. The idea here is to compensate by **brute-force exploration**: run L-BFGS-B from many random starting points and keep the best result. L-BFGS-B does the fast local convergence; random starts do the global search.

With cheap AD gradients: each L-BFGS-B run takes ~20 × 6ms ≈ 120ms. 50 random starts: ~6 seconds total. Trivially parallelizable (independent runs). Provides a strong baseline and is implementable in one day by reusing the existing local solver.

### Proposal B: Mini-batch Adam → L-BFGS-B polish

Two phases with distinct roles. **Phase 1 (Adam with noise): global exploration.** Sub-sampling observation points creates stochastic gradients that prevent convergence to any single local minimum — the optimiser keeps moving through the landscape. **Phase 2 (L-BFGS-B): local polish.** Once Phase 1 has found a promising basin, L-BFGS-B with the full dataset converges precisely to the bottom of that basin.

This scales better to many parameters than multi-start (Proposal A), because Adam explores the landscape continuously rather than sampling it at discrete random points. In high dimensions, random starts become sparse and may miss narrow basins; gradient-guided exploration follows the landscape structure.

Sub-sample observation points to create gradient noise, run Adam, then polish:

```
p = random initial parameters
for step in range(max_steps):
    batch_idx = random.sample(range(T), size=B)
    J_batch, grad_batch = evaluate(p, obs_times[batch_idx])
    p = adam_step(p, grad_batch, step)
    B = min(B + 1, T)  # annealing

p = lbfgsb(p, full_cost, full_gradient)  # polish
```

Advantages over basin-hopping: no problem-specific perturbation tuning; noise scale adapts to landscape curvature automatically. Natural extension to posterior sampling (SGLD) for uncertainty quantification.

### Parallelisation

AD-based optimisation has two natural levels of parallelism:

**Batch evaluation (within one optimiser run).** The recorded AD tape can evaluate cost and gradient at multiple parameter vectors simultaneously using thread-level parallelism. We tested this with the implicit Euler + IFT solver (PR #251): batch of 100 evaluations in 36ms (1 thread), 23ms (4 threads).

**Independent runs (across optimiser instances).** Multi-start L-BFGS-B is embarrassingly parallel: each start is independent. Mini-batch Adam is sequential within one run, but multiple runs from different starting points parallelise the same way.

The two levels compose: N parallel optimiser instances, each using M threads for batch gradient evaluation. For a machine with C cores: N × M ≈ C.

### Recommendation

1. **Start with multi-start L-BFGS-B** (Proposal A) — one day to implement, provides baseline
2. **Then try mini-batch Adam** (Proposal B) — gradient-guided exploration scales better to many parameters; natural connection to uncertainty quantification via SGLD
3. If local minima persist, consider **multiple shooting** (Method 5) — addresses the root cause by smoothing the landscape

Happy to contribute code for either approach.

### References

- Wales & Doye (1997). Global Optimization by Basin-Hopping. J. Phys. Chem. A.
- Welling & Teh (2011). Bayesian Learning via Stochastic Gradient Langevin Dynamics. ICML.
- Raginsky, Rakhlin, Telgarsky (2017). Non-convex Learning via SGLD. COLT.
- Gelfand & Mitter (1991). Recursive Stochastic Algorithms for Global Optimization. SIAM J. Control Optim.
- Bock & Plitt (1984). A Multiple Shooting Algorithm for Direct Solution of Optimal Control Problems. IFAC.
- Rinnooy Kan & Timmer (1987). Stochastic Global Optimization Methods. Math. Programming.

**Comment by FinbarArgus** [comment]
This is another good option:

https://github.com/serdarozkan/TikTak

**Comment by FinbarArgus** [comment]
@lakshtanov I am going to start implementing proposal A and B. I'll let you know how it goes.
